flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/PartitionITCase.java

```diff
@@ -546,43 +549,264 @@ public void testRangePartitionInIteration() throws Exception {
+			return first;
+		}
+
+		public void setFirst(Long first) {
+			this.first = first;
+		}
+
+		public Long getSecond() {
+			return second;
+		}
+
+		public void setSecond(Long second) {
+			this.second = second;
+		}
+
+		public ComparablePojo(Long first,
+							  Long second) {
+			this.first = first;
+			this.second = second;
+		}
+
+		public ComparablePojo() {
+		}
+
+		@Override
+		public int compareTo(ComparablePojo o) {
+			final int firstResult = Long.compare(this.first, o.first);
+			if (firstResult == 0) {
+				return (-1) * Long.compare(this.second, o.second);
+			}
+
+			return firstResult;
+		}
+	}
+
 	private static class ObjectSelfKeySelector implements KeySelector<Long, Long> {
 		@Override
 		public Long getKey(Long value) throws Exception {
 			return value;
 		}
 	}
 
-	private static class MinMaxSelector implements MapPartitionFunction<Long, Tuple2<Long, Long>> {
+	private static class MinMaxSelector<T> implements MapPartitionFunction<T, Tuple2<T, T>> {
+
+		private final Comparator<T> comparator;
+
+		public MinMaxSelector(Comparator<T> comparator) {
+			this.comparator = comparator;
+		}
+
 		@Override
-		public void mapPartition(Iterable<Long> values, Collector<Tuple2<Long, Long>> out) throws Exception {
-			long max = Long.MIN_VALUE;
-			long min = Long.MAX_VALUE;
-			for (long value : values) {
-				if (value > max) {
+		public void mapPartition(Iterable<T> values, Collector<Tuple2<T, T>> out) throws Exception {
+			Iterator<T> itr = values.iterator();
+			T min = itr.next();
+			T max = min;
+			T value;
+			while (itr.hasNext()) {
+				value= itr.next();
+				if (comparator.compare(value, min) < 0) {
+					min = value;
+				}
+				if (comparator.compare(value, max) > 0) {
 					max = value;
 				}
 
-				if (value < min) {
-					min = value;
-				}
 			}
-			Tuple2<Long, Long> result = new Tuple2<>(min, max);
+
+			Tuple2<T, T> result = new Tuple2<>(min, max);
 			out.collect(result);
 		}
 	}
 
-	private static class Tuple2Comparator implements Comparator<Tuple2<Long, Long>> {
+	private static class Tuple2Comparator<T> implements Comparator<Tuple2<T, T>>, Serializable {
+
+		private final Comparator<T> firstComparator;
+		private final Comparator<T> secondComparator;
+
+		public Tuple2Comparator(Comparator<T> comparator) {
+			this(comparator, comparator);
+		}
+
+		public Tuple2Comparator(Comparator<T> firstComparator,
+								Comparator<T> secondComparator) {
+			this.firstComparator = firstComparator;
+			this.secondComparator = secondComparator;
+		}
+
 		@Override
-		public int compare(Tuple2<Long, Long> first, Tuple2<Long, Long> second) {
-			long result = first.f0 - second.f0;
+		public int compare(Tuple2<T, T> first, Tuple2<T, T> second) {
+			long result = firstComparator.compare(first.f0, second.f0);
 			if (result > 0) {
 				return 1;
 			} else if (result < 0) {
 				return -1;
 			}
 
-			result = first.f1 - second.f1;
+			result = secondComparator.compare(first.f1, second.f1);
```

Comment:
I think we should only order on the min-element of the records (`first.f0` and `second.f0`). If `first.f0 == `second.f0`, the test must fail, because it would indicate that the same record ended up in two partitions, which is not allowed.